### ISSUE TITLE
PyThaiNLP 2.1 change log

### ISSUE THREAD

**Issue by wannaphongcom** [PyThaiNLP/pythainlp#181]
[WORK IN PROGRESS - Tentative release date: 10 Dec 2019]

## Corpus
- Add Thai female, male names corpus (#217 #297) - thanks @korkeatw @c4n @bact 
  - Thai male, female names corpus https://github.com/korkeatw/thai-names-corpus
  - General Election 2019 candidate names https://github.com/codeforthailand/dataset-election-62-candidates/tree/master/data
- Add `PYTHAINLP_DATA_DIR` environment variable to set location of downloaded data (default is `~/pythainlp-data`) (#238 #294) - thanks @dhpollack @abhabongse 
- Remove racing condition when create data directory (#278) - thanks @abhabongse 

## Localization
-  Add `pythainlp.util.thai_time` Time spell out to Thai words (#303) thanks @wannaphongcom @abhabongse @bact 

## Tokenizer
- `pythainlp.tokenize.Tokenizer` is now immediately available when `import pythainlp` (https://github.com/PyThaiNLP/pythainlp/commit/79432c2c29718d0531c8aaca862e592bed995bbf) - thanks @korakot 
- Add [`ssg`](https://github.com/ponrawee/ssg), a CRF syllable segmentor (#229 #237 #242) - thanks @wannaphongcom @ponrawee @heytitle 
- Add [`AttaCut`](https://github.com/pythainlp/attacut), a fast and accurate tokenizer, is now available through `engine="attacut"` in `pythainlp.tokenize.word_tokenize()` (#258, #261) - thanks @heytitle @bkktimber 
- Tokenization benchmark (#248 #268 #269) - thanks @wannaphongcom @heytitle
- New engine `newmm-safe` for `pythainlp.tokenize.word_tokenize()` - a `newmm` engine with additional mechanism to avoid possible exponentially long wait for long text with a lot of ambiguity in breaking points. (#302) - thanks @bact

## Spellchecker
- Avoid the spell check for numeric string (#276 #288) - thanks @nawaphonOHM @Peradon 

## Named-Entity Tagger
- Add html-like tag in output (#262) - thanks @wannaphongcom 

## Dependency cleaning
Removing and updating many dependencies - thanks @c4n @artificiala @cstorm125 @korakot @bact @wannaphongcom 

**Remove:**
- `keras`, `tensorflow` (#202 #235 #246) - Thai romanization is now implemented in PyTorch 
-  `fastai` (#252) - removing and replacing `pythainlp.ulmfit` preprocessing-related code with a self-implemented one
- `marisa-trie` (#277) - removing and replacing with native Trie implementation
-  `deepcut` (#283) - removing, word tokenizer still support `engine="deepcut"` but the user needs to install dependencies (`deepcut`, `keras`, `tensorflow`) by themselves

**Update:**
- `artagger` (#281) - updating to use one from the main repo (was depends on a fork)
- Include only direct dependencies in `setup.py` (#275)
- Push the version requirement for dependencies to the lowest possible (#292)

## Documentation
- Docstring and type annotation fixes (#279) - thanks @abhabongse  
- Updated tutorial notebooks and moved to https://github.com/PyThaiNLP/tutorials (#282) - thanks @artificiala @cstorm125 
- Citation fix (#284) - thanks @heytitle
- Docstring code and output example style changed, make it easier to copy & paste code (#293) - thanks @heytitle

## Others
- Fix normalization, to include case THANTHAKHAT and SARA U, SARA UU (#244) - thanks @korakot @ekapolc
- Better command-line interface (#251, #271) - thanks @heytitle @wannaphongcom 
- Improve code readability (#287 #290 #291) - thanks @boomsquared @Peradon
- Refactor:
  - Refactor the test files (#234) - thanks @artificiala
  - Optimize keyboard layout switching translation code & digits translation (#280) - thanks @abhabongse
  - Refactor util package as well as improve performance (#295) - thanks @Peradon

**Comment by heytitle** [comment]
it would be more efficient if we have a way to automatically generate this change log.

**Comment by cstorm125** [comment]
@heytitle agreed. do you have any suggestion?

**Comment by bact** [comment]
Publicly announced the 2.1, close the issue.
https://www.blognone.com/node/113587
